### PR TITLE
Spectrogram UX improvements

### PR DESCRIPTION
**Pull request by Brollyy**
#24

Added a way to resize the column with spectrogram to get both a better look at the image and to allow users to drag it closer to the note editor to check their alignment.

Added caching for spectrogram to make sure we don't re-do expensive calculations unnecessarily when opening a map again or resizing the spectrogram.

Added user settings for customizing the spectrogram, including algorithm used, frequency range and colormap (previously it was always MelScale, 16kHZ, Blues). I've set the defaults to Standard, 11kHz, which seems about the same settings that are used by MMA2 for their spectrogram. I've left Blues as the default colormap for theming, but after testing a few I actually found I prefer Lopora more and I saw that MMA2 uses Plasma.

image

White bar allows user to change the width of the spectrogram by dragging it:
image

New section in user settings (only displays after user enables spectrogram):
image

Using Standard with 11kHz takes a little longer to load initially, but provides much higher resolution, especially with the ability to resize:
image

### CODE REVIEW

**Comment by Brollyy** [comment]
I'll try to setup a call with other mappers to see if they have any other UX improvements in mind after using the Beta.

In terms of the spectrogram content itself, I've tried fiddling with the options a bit, but honestly I don't really understand enough of what goes under the hood technically to know how to improve the readability of it.

File: Classes/EddaConstants.cs
Comment (Brollyy) [inline Classes/EddaConstants.cs:158]: I'm not sure if we should also make this available for users when the type is MelScale. 

While increasing it will slightly improve the resolution, it also has a large impact on the contents, as it's basically a logarithmic scale - most of the added pixels are spent on very low (below 100Hz), uninteresting frequencies.

File: Windows/SettingsWindow.xaml
Comment (Brollyy) [inline Windows/SettingsWindow.xaml:183]: I've discussed having this kind of warning with Waatsu, as he tends to package his maps outside of Edda's Export Map function and this is also shown in his mapping tutorial (it's quite outdated at the moment).

If you have any good idea about a place to put the cached images without risking an impact of those being included in the ZIP files by mistake, I'll gladly change the save location and remove this warning.

File: Windows/SettingsWindow.xaml
Comment (Brollyy) [inline Windows/SettingsWindow.xaml:23]: I've added scrolling for the settings, as the new section was quite long and it didn't make sense to make the window even longer.

File: Windows/SettingsWindow.xaml
Comment (PKBeam) [reply Windows/SettingsWindow.xaml:184]: We could put them in a temp folder in AppData, but we already store backups in the map folder so adding a few more images isn't really going to change anything.

Maybe we could move backups out of the map folder in future and the spectrograms could go with them.

File: Classes/EddaConstants.cs
Comment (PKBeam) [reply Classes/EddaConstants.cs:160]: this is probably something we should hide from the user.

we can tune this value ourselves for quality/performance and it should be able to work well for any song.

**Review by PKBeam — APPROVED** [review]
approved - i'll assume you've tested all these changes.

**Comment by Brollyy** [comment]
> 

I still have to test some edge cases for performance (e.g. rapidly changing settings) and waiting on feedback from Waatsu after the last change with the flip. Will let you know once it's ready to be merged.

File: Classes/Audio/VorbisSpectrogramGenerator.cs
Comment (Brollyy) [inline Classes/Audio/VorbisSpectrogramGenerator.cs:133]: With current settings, the cached images are quite large for longer songs (~9 MB) and since all selected possibilities are cached, the storage racks up quickly.

Investigate if reducing the horizontal resolution still gives ok results - maybe add an option for the user to reduce quality to speed up the generation and reduce storage space?

File: Classes/Audio/VorbisSpectrogramGenerator.cs
Comment (Brollyy) [inline Classes/Audio/VorbisSpectrogramGenerator.cs:43]: This should probably cancel previous drawing process - now it's possible to change the settings multiple times in a row while the previous image is still being generated and get the old image saved under new filename.

File: Classes/Audio/VorbisSpectrogramGenerator.cs
Comment (Brollyy) [reply Classes/Audio/VorbisSpectrogramGenerator.cs:143]: Previous step size gives very good quality of the image - I've set that as High quality.

Increasing the step size by a factor of 2 reduced the quality only slightly, almost not noticeably - I've set that as Medium quality and a default option.

Increasing the step size by a factor of 4 reduced the quality significantly - I've set that as Low quality option.

File: Classes/EddaConstants.cs
Comment (Brollyy) [reply Classes/EddaConstants.cs:162]: Values over 100 don't increase the quality significantly, but cause the logarithmic drift I've described above:

100:
image

150:
image

200:
image

**Review by Brollyy — COMMENTED** [review]
There's an issue where with some specific songs and settings, spectrogram doesn't render any image at all - there's a new file created in cache, but it's empty.

EDIT: there's a generic GDI+ exception when saving the file. Looking at potential list of causes (https://stackoverflow.com/questions/1053052/a-generic-error-occurred-in-gdi-jpeg-image-to-memorystream) I was able to narrow it down to this one: **GDI+ limits the maximum height of an image to 65500**

Currently, this limit is exceeded on longer songs:
image

**Comment by Nytilde** [comment]
it's no good practise to add something lager than one line within a try-catch code block. 
just as my recommendation: i would either outsource those parts into a function or apply try-catch only to the part where the error can actually occur.

**Comment by Brollyy** [comment]
I've implemented option to load the spectrogram in chunks to support longer maps and done some code refactoring for readability in VorbisSpectrogramGenerator.

I've tested all of those changes myself, so it should be fine to merge if you don't see any issues - if anything comes up during beta that I didn't find, I'll fix it then.

**Pull request merged**
